File: LeetCode/CodingInterviews/Go/verifyPostOrder.go
package coding

func verifyPostorder(postorder []int) bool {
	if len(postorder) == 0 {
		return true
	}
	var check func(int, int) bool
	check = func(left, right int) bool {
		if left > right {
			return true
		}
		var root = postorder[right]
		var k = left
		for k < right && postorder[k] < root {
			k++
		}
		for i := k; i < right; i++ {
			if postorder[i] < root {
				return false
			}
		}
		return check(left, k-1) && check(k, right-1)
	}
	return check(0, len(postorder)-1)
}

File: LeetCode/CodingInterviews/C/VerifyPostOrder.cpp
#include "main.h"

class Solution {
  public:
    bool verifyPostorder(vector<int> &postorder) {
        if (postorder.empty())
            return false;
        return check(postorder, 0, postorder.size() - 1);
    }

  private:
    bool check(vector<int> &post, int left, int right) {
        if (left >= right)
            return true;
        int root = post[right];
        int k = left;
        while (k < right && post[k] < root)
            ++k;
        for (int i = k; i < right; i++)
            if (post[i] < root)
                return false;
        return check(post, left, k - 1) && check(post, k, right - 1);
    }
};